domain/permission/service/service.go

```diff
@@ -17,15 +18,17 @@ type State interface {
 	// CreatePermission gives the user access per the provided spec.
 	// It requires the user/target combination has not already been
 	// created.
-	CreatePermission(ctx context.Context, spec UserAccessSpec) (permission.UserAccess, error)
+	CreatePermission(ctx context.Context, spec state.UserAccessSpec) (permission.UserAccess, error)
 
 	// DeletePermission removes the given subject's (user) access to the
 	// given target.
 	DeletePermission(ctx context.Context, subject string, target permission.ID) error
 
-	// UpdatePermission updates the subject's (user) access to the given
-	// target to the given access.
-	UpdatePermission(ctx context.Context, subject string, target permission.ID, access permission.Access) error
+	// UpdateMaybeCreatePermission updates the permission on the target for the given
+	// subject (user). The api user must have Admin permission on the target. If a
+	// subject does not exist, it is created using the subject and api user. Access
+	// can be granted or revoked.
+	UpdateMaybeCreatePermission(ctx context.Context, args state.UpdateMaybeCreatePermissionArgs) error
```

Comment:
Upsert is the common name for this action.

Reply:
done